src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -58,6 +61,35 @@ public partial class Activity
         /// </summary>
         public string ParentId { get; private set; }
 
+
+        /// <summary>
+        /// Root Id is substring from Activity.Id (or ParentId) between '/' (or beginning) and first '.'.
+        /// Filtering by root Id allows to find all Activities involved in operation processing.
+        /// RootId may be null if Activity has neither ParentId nor Id.
+        /// See <see href="https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md#activity-id"/> for more details
+        /// </summary>
+        public string RootId
+        {
+            get
+            {
+                //we expect RootId to be requested at any time after activity is created, 
+                //possibly even before it was started for sampling or logging purposes
+                //Presumably, it will be called by logging systems for every log record, so we cache it.
+                if (_rootId == null)
+                {
+                    if (Id != null)
+                    {
+                        _rootId = getRootId(Id);
+                    }
+                    else if (ParentId != null)
```

Comment:
I don't think we need to check for ParentId.   We really try hard to set a sensible Id and thus ID can only be null in the 'wacky' case where we have not called Start(), and the fact that RootId returns null in that case is no problem.  

On the other hand, it does not hurt (just code complexity (and explanation complexity)  I would drop it but if you want to keep it I guess that is OK too.